datasource.go

```diff
@@ -0,0 +1,204 @@
+		}
+	}()
+
+	refId := req.Queries[0].RefId
+	modelJson, err := simplejson.NewJson([]byte(req.Queries[0].ModelJson))
+	if err != nil {
+		return nil, err
+	}
+
+	query := modelJson.Get("rawQuery").MustString()
+	request, err := createRequest(req, query)
+	if err != nil {
+		return nil, err
+	}
+
+	response, err := ctxhttp.Do(ctx, httpClient, request)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("invalid status code. status: %v", response.Status)
+	}
+
+	defer response.Body.Close()
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		return nil, err
+	}
+
+	return parseResponse(body, refId)
+}
+
+func createRequest(req *datasource.DatasourceRequest, query string) (*http.Request, error) {
+	body := ""
+	method := http.MethodGet
+	headers := http.Header{}
+	url, err := url.Parse(req.Datasource.Url)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse clickhouse url: %w", err)
+	}
+
+	params := url.Query()
+	params.Add("query", query+" FORMAT JSON")
+
+	// TODO Fix basic authorization. We have access to basicAuthPassword but not
+	// the basic auth name. Users will have to use the useYandexCloudAuthorization
+	// option instead for clickhouse auth.
+	secureOptions := req.Datasource.DecryptedSecureJsonData
+	options := make(map[string]interface{})
+	err = json.Unmarshal([]byte(req.Datasource.JsonData), &options)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse clickhouse options: %w", err)
+	}
+
+	for k, v := range options {
+		switch k {
+		case "usePOST":
+			method = http.MethodPost
+			params.Del("query")
+			body = query
+			break
+		case "defaultDatabase":
+			db, _ := v.(string)
+			params.Add("database", db)
+			break
+		case "addCorsHeaders":
+			params.Add("add_http_cors_header", "1")
+			break
+		case "useYandexCloudAuthorization":
+			chUser := ""
+			chKey := ""
+
+			if user, ok := options["xHeaderUser"]; ok {
+				chUser, _ = user.(string)
+			}
+
+			if key, ok := secureOptions["xHeaderKey"]; ok {
+				chKey = key
+			}
+
+			headers.Add("X-ClickHouse-User", chUser)
+			headers.Add("X-ClickHouse-Key", chKey)
+			break
+		default:
+			if strings.HasPrefix(k, "httpHeaderName") {
+				headerKey := strings.Replace(k, "Name", "Value", 1)
+				value := ""
+				name, _ := v.(string)
+				if hv, ok := secureOptions[headerKey]; ok {
+					value = hv
+				}
+
+				return nil, fmt.Errorf("parse key %s", k)
+
+				headers.Add(name, value)
+			}
+			break
+		}
+	}
+
+	url.RawQuery = params.Encode()
+	request, err := http.NewRequest(method, url.String(), bytes.NewBufferString(body))
+	if err != nil {
+		return nil, err
+	}
+
+	request.Header = headers
+	return request, nil
+}
+
+func parseResponse(body []byte, refId string) (*datasource.DatasourceResponse, error) {
+	parsedBody := ClickHouseResponse{}
+	err := json.Unmarshal(body, &parsedBody)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse response json: %s: %w", body, err)
```

Comment:
`body` is a byte and must be converted to string before printing

Reply:
@hagen1778 please look at https://play.golang.org/p/l1nRLOTQOV7 
[]byte will convert to string automatically
also int, real etc.